### PR TITLE
[buildingplan] step3: construct buildings from lua 

### PR DESCRIPTION
**Pull request by myk002**
#1640
Replace C++ building construction code with lua constructBuilding so we can get the proper job_item filters set. these filters will be used when we replace the core buildingplan algorithm in the next PR.

Code can be reviewed after #1673 is merged

### CODE REVIEW

**Comment by myk002** [comment]
ready for review

**Review by lethosor — COMMENTED** [review]
Forgot to submit, oops

File: plugins/buildingplan.cpp [hunk 1]
```diff
@@ -287,6 +287,33 @@ static bool is_planmode_enabled(BuildingTypeKey key)
...
+
+    if (!(lua_checkstack(L, 1) &&
+          Lua::PushModulePublic(out, L, "plugins.buildingplan",
+                                "construct_building_from_ui_state") &&
+          Lua::SafeCall(out, L, 0, 1)))
+    {
+        return false;
+    }
+
+    auto bld = static_cast<df::building *>(LuaWrapper::get_object_ref(L, -1));
```
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/buildingplan.cpp]: done. thanks!

File: plugins/fortplan.cpp [hunk 2]
```diff
@@ -47,8 +50,38 @@ struct BuildingInfo {
...
+        Lua::Push(L, type);
+        Lua::Push(L, cursor.x);
+        Lua::Push(L, cursor.y);
+        Lua::Push(L, cursor.z);
+
+        if (!Lua::SafeCall(out, L, 4, 1))
+            return false;
+
+        auto bld =
+            static_cast<df::building *>(LuaWrapper::get_object_ref(L, -1));
```
[the remark this answers is not in the document]
Comment (myk002) [reply plugins/fortplan.cpp]: done

**Comment by lethosor** [comment]
Evidently I should have tested my CheckDFObject suggestion sooner. It turns out that because it's running outside of protected mode, it will end up calling abort() on error:
```
PANIC: unprotected error in call to Lua API (at index -1: invalid pointer type; expected: building)
```
Note that `nil` fails silently, which is fine. This only occurs if the Lua function returns a non-nil and non-building pointer, which is pretty unlikely.

I'm playing around with error handling already so I'll come up with something, but this looks good!

**Review by lethosor — APPROVED** [review]

**Pull request merged**